unittests/chain_tests.cpp

```diff
@@ -0,0 +1,77 @@
+#include <eosio/chain/controller.hpp>
+#include <eosio/chain/global_property_object.hpp>
+#include <eosio/chain/permission_object.hpp>
+#include <eosio/chain/resource_limits.hpp>
+#include <boost/test/unit_test.hpp>
+#include <eosio/testing/tester.hpp>
+
+#include "fork_test_utilities.hpp"
+
+using namespace eosio;
+using namespace eosio::chain;
+using namespace eosio::testing;
+
+BOOST_AUTO_TEST_SUITE(chain_tests)
+
+BOOST_AUTO_TEST_CASE( replace_producer_keys ) try {
+   validating_tester tester;
+
+   const auto head_ptr = tester.control->head_block_state();
+   BOOST_ASSERT(head_ptr);
+
+   const auto new_key = get_public_key(name("newkey"), "active");
+
+   // make sure new keys is not used
+   for(const auto& prod : head_ptr->active_schedule.producers) {
+      for(const auto& key : prod.authority.get<block_signing_authority_v0>().keys){
+         BOOST_ASSERT(key.key != new_key);
+      }
+   }
+
+   const auto old_version = head_ptr->pending_schedule.schedule.version;
+   BOOST_REQUIRE_NO_THROW(tester.control->replace_producer_keys(new_key));
+   const auto new_version = head_ptr->pending_schedule.schedule.version;
+   // make sure version not been changed
+   BOOST_ASSERT(old_version == new_version);
+
+   const auto& gpo = tester.control->db().get<global_property_object>();
+   BOOST_ASSERT(!gpo.proposed_schedule_block_num);
+   BOOST_ASSERT(gpo.proposed_schedule.version == 0);
+   BOOST_ASSERT(gpo.proposed_schedule.producers.empty());
+
+   const uint32_t expected_threshold = 1;
+   const weight_type expected_key_weight = 1;
+   for(const auto& prod : head_ptr->active_schedule.producers) {
+      BOOST_REQUIRE_EQUAL(prod.authority.get<block_signing_authority_v0>().threshold, expected_threshold);
+      for(const auto& key : prod.authority.get<block_signing_authority_v0>().keys){
+         BOOST_REQUIRE_EQUAL(key.key, new_key);
+         BOOST_REQUIRE_EQUAL(key.weight, expected_key_weight);
+       }
+   }
+} FC_LOG_AND_RETHROW()
+
+BOOST_AUTO_TEST_CASE( replace_account_keys ) try {
+   validating_tester tester;
+   const name usr("eosio");
+   const name active_permission("active");
+   const auto& rlm = tester.control->get_resource_limits_manager();
+   const auto* perm = tester.control->db().find<permission_object, by_owner>(boost::make_tuple(usr, active_permission));
+   BOOST_ASSERT(perm != NULL);
+
+   const int64_t old_size = (int64_t)(chain::config::billable_size_v<permission_object> + perm->auth.get_billable_size());
```

Comment:
Why the cast to `int64_t` here and below?

Reply:
it is used in replace_account_keys  implementation, https://github.com/EOSIO/eos/pull/8929/files

Reply:
Yuck, but ok.